### PR TITLE
Separate large perf benchmarks into their own legs

### PR DESCRIPTION
**Pull request by adiaaida**

### CODE REVIEW

**Comment by adiaaida** [comment]
@dotnet-bot test Perf Build and Test please

**Comment by adiaaida** [comment]
@dotnet-bot test Perf Build and Test please

**Comment by adiaaida** [comment]
@dotnet-bot test Perf Build and Test please

**Comment by adiaaida** [comment]
@dotnet-bot test Perf Build and Test please

**Comment by adiaaida** [comment]
@dotnet-bot test Perf Build and Test please

**Comment by adiaaida** [comment]
@dotnet-bot test Perf Build and Test please

**Comment by adiaaida** [comment]
@dotnet-bot test Perf Build and Test please

**Comment by adiaaida** [comment]
@dotnet-bot test Perf Build and Test please

**Comment by adiaaida** [comment]
@dotnet-bot test Perf Build and Test please

**Comment by adiaaida** [comment]
@dotnet-bot test Perf Build and Test please

**Comment by adiaaida** [comment]
@dotnet/rap-team With this change, the runtime for the perf leg is down to 1hr. It's still about 10 minutes longer than the slowest functional test, but part of that, I think, is the creation of VMs. The slowest testing leg is the x86 Math test, which takes about 21 minutes total (12 of which are running the test). There doesn't seem to be a way to split that one any further, so it is our limiting factor.

/cc: @valenis

File: buildpipeline/perf-pipeline.groovy [hunk 1]
```diff
@@ -19,25 +32,32 @@ def windowsBuild(String arch, String config, String pgo, boolean isBaseline) {
     }
 
     bat "set __TestIntermediateDir=int&&.\\build.cmd -${config} -${arch} -skipbuildpackages ${pgoBuildFlag}"
+    bat "tests\\runtest.cmd ${config} ${arch} GenerateLayoutOnly"
+    bat "rd /s /q bin\\obj"
 
     // Stash build artifacts. Stash tests in an additional stash to be used by Linux test runs
-    stash name: "nt-${arch}-${pgo}${baselineString}-build-artifacts", includes: 'bin/**'
+    stash name: "nt-${arch}-${pgo}${baselineString}-build-artifacts", includes: 'bin/'
```
Comment (DrewScoggins) [inline buildpipeline/perf-pipeline.groovy]: Why did we get rid of the recursive stash of the bin directory here? Was it no longer needed as we had no other directories under bin, or is this part of the size reduction?

Comment (adiaaida) [reply buildpipeline/perf-pipeline.groovy]: That's a mistake

**Comment by brianrob** [comment]
@adiaaida, I see that the benchmark names are hardcoded in the groovy file.  What happens if I add a new benchmark directory to CoreCLR?

**Comment by adiaaida** [comment]
@brianrob we'd have to decide how it should be run (as a suite/by itself) and add it. This definitely isn't a perfect solution.

**Comment by brianrob** [comment]
Does that mean that by default it won't run?

**Comment by adiaaida** [comment]
Yes

**Comment by brianrob** [comment]
Hmm... I think we need to have a solution for this.  This means that adding performance tests is no longer easy, and requires intervention from our team.

As a first cut I think that we should make sure that any unaccounted-for tests get to run, perhaps in their own leg.  Then, I think we should look at ways to partition things more generically.

I get worried having hardcoded stuff because it means that we've added more surface that we're going to have to manage.

**Comment by adiaaida** [comment]
I agree, but I don't think that the groovy code will know about anything to do with the source code structure, meaning, I can't have it generate legs based off of a directory structure because it doesn't know about it. One way we might be able to handle some of this is change the directory structure of the Performance directory. We'd move some of the folders we currently have into different folders (for the split of tests that should be run more less together), and then new tests could be added there, instead of in the base CodeQuality directory.

**Comment by adiaaida** [comment]
@dotnet-bot test Perf Build and Test please

**Comment by brianrob** [comment]
Yup, makes sense that Jenkins can't do this, but what if Jenkins just created N jobs, and then those jobs decided what they are responsible for?  If that's the case, then they have full access to the repo and can ensure that all tests get run.  You can start with N == whatever N currently is.  The other nice thing about this is that if we want to make changes to how things are load balanced it's easy to do so and will immediately be tested by the CI for the PR that makes the changes as opposed to jumping through the hoops required to change the groovy files (unless that's gotten easier since I last did this).

It's reasonable to have hints built-in that specify what tests are slower or faster so that things get grouped together in a sane way, but I think it's important that we parallelize in a way that doesn't require us to opt new tests in with a change to the perf.groovy.  This just sounds like another thing to manage...

What do you think?

**Comment by adiaaida** [comment]
So you're saying do something like create N legs, and then have a for loop in the main batch script of the leg loop over the directories of the Jit\Performance\CodeQuality directory with an incrementer, and when we get to the leg number, run that particular benchmark suite, and on the N'th leg, we just run whatever is left? That sort of thing? I think it could potentially work, but that it wouldn't load balance as nicely as we have now because the suites aren't ordered in terms of how long they are, so we could end up with the last leg having too many benchmarks, or too many long benchmarks trying to run together and slowing us down.

I also think that would take quite a bit of experimentation to get right, and @valenis suggested we wanted to get these legs back online ASAP, so if we go down that route, it may need to be a second step after we can get something in.

**Comment by adiaaida** [comment]
For what it's worth, I'd like to point out that we already have to do this sort of manual adding of the tests when we add new scenario benchmarks, and we already have the PerfLab test hard-coded in the groovy scripting because it's run differently (and is in a different directory). I do understand not wanting to add addition work for when adding tests, and I agree that we should have a more automatic system for discovering tests, I just wanted to point out that we already do do this.

**Comment by brianrob** [comment]
@adiaaida thanks for the explanation.  I didn't quite understand at first, but @DrewScoggins came by to chat and explained it to me.  I think in general, we should try and have things as automated as possible, but there's certainly a line where the benefit outweighs the cost.  We could very quickly get to this point if we try to automate too much here.

I think what you have here - using the test names in the perf.groovy is an OK first cut so that we can reduce time and start to see the benefits of getting the per-commit data.  The next two things that I think we should do (both of these before we enable per-PR runs) are:

 - Switch to a scheme where the determination of what runs in each test slice is determined by the slice itself (I'll explain below how I envision this).
 - Do a survey of past failures of the job and see if we need to do any hardening for reliability in addition to reducing job execution time.

As far as the scheme I allude to above, I get concerned that we're in a situation where it's technically possible to add benchmarks and not have them get picked up by the perf jobs.  I realize that it already happens today, but we should fix that.  My proposal for how to fix that and the load balancing is as follows:

1. Have the perf pipeline job spawn N jobs, each of which will execute a slice of total set of tests.  Each job knows its slice number (its passed in as a parameter).
2. Check-in a configuration file into the repo that has mapping from slice number to the set of tests to actually run.
3. Have the last slice determine if there are any other tests that aren't accounted for by the configuration file, and have it include them in its run.
4. On PR runs, have a final job that determines if any perf test changes were made, and if so, do some additional checks.  These checks can include figuring out if any one leg outlasts the others significantly, or if any one leg lasts for longer than a specified period of time.  This will help to catch imbalances and fast/slow creep infrastructure issues such as that stash/unstash take too long.  These can be rectified by adding the new test to another leg, increasing the timeout, rebalancing, etc. but will be caught before the changes are merged.  Then we get notified for help.

This ensures that all tests get run, and that if things start to go out of "compliance" jobs fail before the offending changes are merged.

How does that sound as a plan going forward?  (Sorry for hijacking your PR too...)

**Comment by adiaaida** [comment]
I think I could make something like that happen. I'm not sure if I can transfer the timing information from leg to a following leg, but I should be able to at least get some sort of testing where we make sure we've run the testing on all of the benchmark folders in the CodeQuality folder.

**Comment by brianrob** [comment]
Fair enough.  Let's chat when you're back about the specifics and see if the plan makes sense.

**Comment by adiaaida** [comment]
@brianrob I have opened #15306, #15307, and #15308 to track some of the issues we have discussed.

**Comment by adiaaida** [comment]
@dotnet-bot test Perf test and build please

**Comment by adiaaida** [comment]
@dotnet-bot test Perf build and test please

**Comment by BruceForstall** [comment]
@adiaaida This change is breaking all ARM and ARM64 testing, which uses tests\arm\tests.lst and tests\arm64\tests.lst to determine where to find the tests, which you have moved.

cc @jashook @RussKeldorph 

E.g.: https://ci.dot.net/job/dotnet_coreclr/job/master/job/arm_cross_checked_windows_nt_flow/206/

**Comment by BruceForstall** [comment]
I opened https://github.com/dotnet/coreclr/pull/15387 to fix this.

**Comment by adiaaida** [comment]
@BruceForstall Sorry about that. I didn't realize we hardcoded paths in other files.

**Pull request merged**